pkg/filesystem/filesystem.go

```diff
@@ -304,48 +314,114 @@ func CompareTgzs(fs billy.Filesystem, leftTgzPath string, rightTgzPath string) (
 			isRightEOF = true
 		case leftErr != nil:
 			// ran into unknown error
-			return false, fmt.Errorf("ran into error while trying to read files in %s: %v", leftTgzPath, leftErr)
+			return false, fmt.Errorf("ran into error while trying to read files in left tgz: %v", leftErr)
 		case rightErr != nil:
 			// ran into unknown error
-			return false, fmt.Errorf("ran into error while trying to read files in %s: %v", rightTgzPath, rightErr)
+			return false, fmt.Errorf("ran into error while trying to read files in right tgz: %v", rightErr)
 		}
 		if !isLeftEOF && left.Typeflag == tar.TypeReg {
-			// compute left hash
-			leftHash := sha1.New()
-			if _, err := io.Copy(leftHash, leftTarReader); err != nil {
-				return false, fmt.Errorf("could not compute hash of %s: %v", left.Name, err)
+			switch filepath.Ext(left.Name) {
+			case ".tgz":
+				// collect contents of file into reader
+				var b bytes.Buffer
+				if _, err := b.ReadFrom(leftTarReader); err != nil {
+					return false, fmt.Errorf("could not read contents of %s: %v", left.Name, err)
+				}
+				r, ok := tgzMap[left.Name]
+				if !ok {
+					r = make([]io.Reader, 2)
+				}
+				r[0] = &b
+				tgzMap[left.Name] = r
```

Comment:
Mirrors the same logic we see below with the hashMap, except the map and the value we want to add to the map (e.g. an io.Reader) is different.

Why can't we just compare tgzs or tars here?
-> Because it's not guaranteed that leftTarReader or rightTarReader are reading files in the same order. This is why we collect it into a map whose keys are the filename and whose values are [leftContents, rightContents] and later run the comparison once the values have both been set.